src/main/java/io/micronaut/build/DockerCracMojo.java

```diff
@@ -0,0 +1,206 @@
+        Set<String> tags = getTags();
+        for (String tag : tags) {
+            ImageReference.parse(tag);
+        }
+        copyScripts(RUN_SCRIPT_NAME);
+        File dockerfile = dockerService.loadDockerfileAsResource(DockerfileMojo.DOCKERFILE_CRAC);
+        BuildImageCmd buildImageCmd = dockerService.buildImageCmd()
+                .withDockerfile(dockerfile)
+                .withBuildArg("BASE_IMAGE", getFromImage().orElse(DEFAULT_BASE_IMAGE))
+                .withBuildArg("CHECKPOINT_IMAGE", checkpointContainerId)
+                .withTags(getTags());
+        dockerService.buildImage(buildImageCmd);
+    }
+
+    private void copyScripts(String... scriptNames) throws IOException {
+        List<String> imageClasspathScopes = Arrays.asList(Artifact.SCOPE_COMPILE, Artifact.SCOPE_RUNTIME);
+        String readinessCommand = getReadinessCommand();
+        mavenProject.setArtifactFilter(artifact -> imageClasspathScopes.contains(artifact.getScope()));
+        File target = new File(mavenProject.getBuild().getDirectory(), "scripts");
+        if (!target.exists()) {
+            target.mkdirs();
+        }
+        processScripts(target, s -> s.replace("@READINESS@", readinessCommand).replace("@MAINCLASS@", mainClass), scriptNames);
+    }
+
+    private void processScripts(File target, UnaryOperator<String> replacement, String... scriptNames) throws IOException {
+        for (String script : scriptNames) {
+            File localOverride = new File(mavenProject.getBasedir(), script);
+            try (InputStream resourceAsStream = localOverride.exists() ?
+                    Files.newInputStream(localOverride.toPath()) :
+                    DockerCracMojo.class.getResourceAsStream("/cracScripts/" + script)) {
+                if (resourceAsStream == null) {
+                    throw new IOException("Could not find script " + script);
+                }
+                List<String> tokenized = replaceTokensInTextStream(resourceAsStream, replacement);
+                Path outputPath = target.toPath().resolve(script);
+                writeStringsToFile(tokenized, outputPath);
+                Files.setPosixFilePermissions(outputPath, POSIX_FILE_PERMISSIONS);
+            }
+        }
+    }
+
+    private static List<String> replaceTokensInTextStream(InputStream resourceAsStream, UnaryOperator<String> replacement) throws IOException {
+        return IOUtils.readLines(resourceAsStream, StandardCharsets.UTF_8)
+                .stream()
+                .map(replacement)
+                .collect(Collectors.toList());
+    }
+
+    private static void writeStringsToFile(List<String> tokenized, Path outputPath) throws IOException {
+        try (BufferedWriter out = Files.newBufferedWriter(outputPath, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+            IOUtils.writeLines(tokenized, System.getProperty("line.separator"), out);
+        }
+    }
```

Comment:
There is a better, Maven-ish way to do this, which is using the Maven Resources Plugin API: https://maven.apache.org/shared/maven-filtering/usage.html

You should be able to inject `MavenResourcesFiltering` in the constructor.

Reply:
Ok, so I tried this but failed hard

How do you use that to copy a classpath resource and replace tokens?  I can get it to copy a file, but not extract the files from the classpath of the mojo

Any clues would be handy...

Reply:
Perhaps you can use `MavenReaderFilter` instead. Something like this:

```java
InputStreamReader reader = new InputStreamReader(anInputStream); //classpath resource
MavenReaderFilter readerFilter; //inject it
MavenReaderFilterRequest req = new MavenReaderFilterRequest();
Properties additionalProperties = new Properties();
additionalProperties.setProperty("READINESS", "..." );
req.setFrom(reader);
req.setFiltering(true);
req.setAdditionalProperties(additionalProperties);
String result = IOUtils.toString(readerFilter.filter(req));
IOUtils.write(result, out); //write to file...
```

I got this from https://github.com/apache/maven-filtering/blob/master/src/test/java/org/apache/maven/shared/filtering/DefaultMavenReaderFilterTest.java

Take a look at the other tests in case you see something better.